sdk/python/tests/test_streaming_control_loop.py

```diff
@@ -0,0 +1,184 @@
+        )
+
+        stream_source = KafkaSource(
+            bootstrap_servers="localhost:9094",
+            message_format=ProtoFormat("class.path"),
+            topic="test_topic",
+            event_timestamp_column="ts_col",
+            created_timestamp_column="timestamp",
+        )
+
+        ft1 = FeatureTable(
+            name=self.table_name,
+            features=features,
+            entities=["driver_car_id"],
+            labels={"team": "matchmaking"},
+            batch_source=batch_source,
+            stream_source=stream_source,
+        )
+
+        # Register Feature Table with Core
+        client.apply_feature_table(ft1)
+
+    def _delete_ft(self, client: Client):
+        client.delete_feature_table(self.table_name)
+
+    def test_streaming_job_control_loop(self) -> None:
+        """ Test streaming job control loop logic. """
+
+        core_servicer = MockCoreServicer()
+
+        processes: List[subprocess.Popen] = []
+
+        def _mock_spark_submit(self, *args, **kwargs) -> subprocess.Popen:
+            # We mock StandaloneClusterLauncher.spark_submit to run a dummy process and pretend
+            # that this is a spark structured streaming process. In addition, this implementation
+            # will keep track of launched processes in an array.
+            result = subprocess.Popen(args=["/bin/bash", "-c", "sleep 600"])
+            processes.append(result)
+            return result
+
+        with patch.object(
+            StandaloneClusterLauncher, "spark_submit", new=_mock_spark_submit
+        ), mock_server(
+            core_servicer, add_CoreServiceServicer_to_server
+        ) as core_service_url:
+            client = Client(
+                core_url=core_service_url,
+                serving_url=SERVING_URL,
+                spark_launcher="standalone",
+                spark_home=os.path.dirname(pyspark.__file__),
+            )
+
+            # Run one iteration of the control loop. It should do nothing since we have no
+            # feature tables.
+            ensure_stream_ingestion_jobs(client=client, all_projects=True)
+
+            # No jobs should be running at this point.
+            assert len(client.list_jobs(include_terminated=True)) == 0
+
+            # Now, create a new feature table.
+            self._create_ft(client, self.features_1)
+
+            # Run another iteration of the control loop.
+            ensure_stream_ingestion_jobs(client=client, all_projects=True)
+
+            # We expect a streaming job to be created for the new Feature Table.
+            assert len(client.list_jobs(include_terminated=False)) == 1
+            assert len(processes) == 1
+
+            first_job_id = client.list_jobs(include_terminated=False)[0].get_id()
+
+            # Pretend that the streaming job has terminated for no reason.
+            processes[0].kill()
+
+            # The control loop is expected to notice the killed job and start it again.
```

Comment:
Should we perhaps add one assert here to test that a job has terminated?

Reply:
Hmm not sure what code path would that help test. We just terminated it one line above this so we know it is dead. The assert below checks that there is one terminated and one started job after we ran `ensure_stream_ingestion_jobs`

Reply:
Fair enough.